### PR TITLE
Added in proper prevention of absorbing negative compound amounts

### LINKED ISSUE #3927: Compound clouds should not allow negative compound taking
In this save, the player cell has ~0.9 glucose but only 0.5 glucose capacity.

https://www.dropbox.com/s/d8qrrqqphtn2bw2/glucose_bug.thrivesave?dl=0

The mitigations in https://github.com/Revolutionary-Games/Thrive/pull/3926 prevent this leading to an infinite compound venting glitch, but the root cause should still be fixed. We can then throw an error if we try to take negative compounds from a cloud.

### PR DESCRIPTION
**Pull request by hhyyrylainen**
**Related Issues**

removes the bandaid and fixes #3927

also I couldn't help sneaking in a change Rider automatically suggested as a cleaner code alternative to one line in the AI code.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Review by Oliveriver — COMMENTED** [review]
Code looks fine. This doesn't address the root cause of the bug though, namely that Buckly's cell ended up containing more compounds than it could store. All of the checks added here only prevent that spiralling into a few more glitches.

Still, if that's what you were aiming for then this is good.

File: src/microbe_stage/CompoundBag.cs [hunk 1]
```diff
@@ -55,6 +55,17 @@ public float GetCompoundAmount(Compound compound)
         return amount;
     }
 
+    /// <summary>
+    ///   The space available for a compound of type. If not useful no free space is ever reported
```
[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/CompoundBag.cs:59]: I didn't intend there to be, I could reword this as "The space available for a compound of given type."

File: src/microbe_stage/CompoundBag.cs [hunk 2]
```diff
@@ -55,6 +55,17 @@ public float GetCompoundAmount(Compound compound)
...
+    /// </summary>
+    /// <param name="compound">The compound type to check</param>
+    /// <returns>The free space available</returns>
+    public float GetFreeSpaceForCompound(Compound compound)
+    {
+        // Due to venting not triggering immediately at capacity, we use max here to avoid negative free space
+        return Math.Max(GetCapacityForCompound(compound) - GetCompoundAmount(compound), 0);
```
[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/CompoundBag.cs:66]: I did go through all the references to Capacity and updated all that I think should be. It'd be nice to phase out Capacity being public entirely, but that needs a bit of a refactor (this is kind of needed anyway as the idea for specializing vacuoles for certain compound types has been floating around for a while).

**Comment by hhyyrylainen** [comment]
> Code looks fine. This doesn't address the root cause of the bug though, namely that Buckly's cell ended up containing more compounds than it could store. All of the checks added here only prevent that spiralling into a few more glitches.

That's entirely intended. If it is wanted that the player cell ends up venting useful compounds in situations where they are temporarily over the limit, that's a separate gameplay discussion.
